### PR TITLE
Fix #367: Framework automation detection

### LINKED ISSUE #367: Read package.json to work out what framework is being used, to self-configure port and build-dir even for Angular apps
## Feature idea

 #365 reduces the amount of configuration users need to do for various framework's build dirs. It can't handle angular apps, though, because the build dir is of the form `/dist/<app-name>`. 
 
If we read the package.json to get the name field, we can make a very good guess at the build dir for angular apps. If we're doing that, we could also peek at the start scripts and application dependencies to make a guess at the port. "If the start script runs ng something something, I should listen on 4200," and so on for Vite, Next.js, etc. 
 
## Additional information

See also discussion in #365.

### PR DESCRIPTION
**Pull request by melloware**
## Describe your changes
PR to allow auto-detecting framework and then configuring for that framework if detected in package.json.

## Fixes Issue
Fix #367: Framework automation detection
 
@ia3andy @holly-cummins  this is what I was thinking to start before adding the code to auto-detect and configure.  Let me know any thoughts before I continue...

### CODE REVIEW

File: docs/modules/ROOT/pages/includes/attributes.adoc [hunk 1]
```diff
@@ -1,5 +1,5 @@
-:quarkus-version: 3.0.0.CR2
-:quarkus-quinoa-version: 2.0.2
+:quarkus-version: 3.0.0.Final
```
Comment (shivam-sharma7) [inline docs/modules/ROOT/pages/includes/attributes.adoc:1]:
```suggestion
:quarkus-version: 3.1.0.Final
```

**Review by shivam-sharma7 — CHANGES_REQUESTED** [review]
It should be change!

**Comment by melloware** [comment]
OK I have it working as far as it goes. I don't use Angular so that could use some testing but this is what happens with Vite.

```
2023-06-02 15:40:16,224 INFO  [io.qua.qui.dep.QuinoaProcessor] VITE framework automatically detected from package.json file.
2023-06-02 15:40:16,224 INFO  [io.qua.qui.dep.QuinoaProcessor] VITE framework setting dev server port: 5173
2023-06-02 15:40:16,224 INFO  [io.qua.qui.dep.QuinoaProcessor] VITE framework setting build directory: 'dist'
```

File: deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java [hunk 2]
```diff
@@ -237,6 +247,53 @@ List<HotDeploymentWatchedFileBuildItem> hotDeploymentWatchedFiles(QuinoaConfig q
         return watchedFiles;
     }
 
+    private void autoDetectFramework(Path packageJsonFile, QuinoaConfig quinoaConfig, LaunchModeBuildItem launchMode) {
```
Comment (ia3andy) [inline deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java:250]: Not sure why, but mutating parameters is always bugging me, I am pretty sure you to @melloware as I begin to see you also like clean things :)

So I guess you considered other options, returning a new instance means, having to assign all other fields which might also cause errors, so I follow you on this.

Still, maybe it should be clear in the method name, so I would split this in two methods:
- `detectFramework`
- 'initDefaultConfig(config, framework)`

And make sure all the mutations are as close as possible. I mean no other logic in between beside the empty checks.

WDYT?

**Review by ia3andy — CHANGES_REQUESTED** [review]
This is awesome @melloware I think it's really great improvement. I just added a small comment, but I am pretty sure you will agree with it :)

File: deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java (hunk 2 above)
Comment (melloware) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java:250]: OK let me see if I can rework this a little bit!

Comment (melloware) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java]: Refactored into two distinct methods.

Comment (melloware) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java]: So far in all my testing that has been the case...but I am by **NO** means a Quarkus threading or build expert. 😄

Comment (ia3andy) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java]: @geoand might know about 1 and 2.

Comment (geoand) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java]: I am pretty sure that there is no guarantee that build steps run in the same thread - IIRC we allow to the processing graph to be completed via multiple threads, so classes that contain `@BuildStep` should not contain any state.

**Review by ia3andy — CHANGES_REQUESTED** [review]
@melloware the multi-threading means we need a bit more work for this to be ok :)

Here is what I suggest, **only for those automatic params** (devserverport, bulddir, framework, ...), we don't read from the config anymore, instead we set the value in the `QuinoaDirectoryBuildItem` (either from autodection or from the config) then we use the BuildItem to read.

-> No more mutation, no risk

**Comment by geoand** [comment]
I have not looked at the code in this PR, but in general, if data needs to be shared by between build steps, it should be "carried" by build items.

**Comment by ia3andy** [comment]
@melloware I want to give it a try, it's in my todo list ;-)

**Review by ia3andy — CHANGES_REQUESTED** [review]
Hey @melloware,

We are getting closer, this is a killer new feature!

I found a bug and have a few suggestions to make it perfect:
- You should use `root.getString("name")` to get the app name, currently, I get:
`ANGULAR framework setting build directory: 'dist/"angular-tour-of-heroes"'`
- If possible, reading the json should only happen when the configuration actually needs it
- Let's introduce a new data object to contain the `DetectedFramework(FrameworkType type, JsonObject packageJson)`, this way we don't use apache dep
- Use `FrameworkType.getBuildDirectory(JsonObject packageJson)` which contains the specific Angular logic to read and use the name of the app

**Comment by melloware** [comment]
OK good suggestions!

**Comment by melloware** [comment]
OK so changes made.

1. Don't parse package.json unless all the defaults are in use.
2. new `DetectedFramework` carrier object.
3. `DetectedFramework.getBuildDirectory()` to get the normal or Angular build directory logic
4. `root.getString("name")` now in use for angular dir.
3.

**Comment by ia3andy** [comment]
new bug detected :)
```
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkiverse.quinoa.deployment.devui.QuinoaDevUIProcessor#createCard threw an exception: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at io.quarkiverse.quinoa.deployment.devui.QuinoaDevUIProcessor.createCard(QuinoaDevUIProcessor.java:79)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

	at io.quarkus.builder.Execution.run(Execution.java:123)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
	... 9 more
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:143)
	at io.quarkiverse.quinoa.deployment.devui.QuinoaDevUIProcessor.createCard(QuinoaDevUIProcessor.java:79)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
	```
	
	
Could you rename `getPort()` to `getDevServerPort` in the same go.

**Comment by melloware** [comment]
OK made the changes tested locally and the Dev UI is working for me now.

**Comment by ia3andy** [comment]
@melloware ok I found another bug haha :) (better find them now than later)

When using with vite, vite is detected but  it does not call `dev`. It calls `start` instead.

While doing that, I think we should either change the default codestart to use vite (instead of webpack) or support webpack.
The reason behind is that, when you generate your first app now and you then put your webui, you forget that you need to remove the configured port (before it's was required). With this we can avoid having the port defined in the config when generating the starter app and allow auto detection for any coming app..

**Comment by melloware** [comment]
Ahh yeah I was still requiring the `start` script but it has me thinking it should use the script it finds first either `dev` or `start`?   Because Next and Vite both use `dev` but others user `start` right?

**Comment by ia3andy** [comment]
> Ahh yeah I was still requiring the `start` script but it has me thinking it should use the script it finds first either `dev` or `start`? Because Next and Vite both use `dev` but others user `start` right?

I would suggest a double check, which is to make sure the command defined in the FrameworkType.devScript can auto-detect with a warning explaining that it's better to manually set the dev script when it's different from the framework default wdyt?

**Comment by melloware** [comment]
Agreed that is a good idea.

**Comment by melloware** [comment]
OK this ended up being a lot of changes but I think all for the good.

1. I gather up all scripts for any framework like "start" and "dev" into a `Set`.
2. I check that Set looking for the first script found that matches.  This will make us future proof if we add more frameworks.
3. I store that script to the side when found so we know whether it was "start" or "dev" etc.
4. Refactored the hardcoded "npm start" to "npm run <command>" so it can be dynamic now.


Tested with Vite, NetxJS, and CRA!

**Review by ia3andy — APPROVED** [review]
LGTM

**Pull request merged**